api/saml/parser.py

```diff
@@ -509,24 +511,34 @@ def _parse_attribute_values(self, attribute_values):
             else:
                 parsed_attribute_values.append(attribute_value)
 
-        return parsed_attribute_values
+        return name_id, parsed_attribute_values
 
-    def _parse_attribute_statement(self, attributes):
+    def _parse_attributes(self, attributes):
+        """Parses SAML attributes
+
+        :param attributes: Dictionary containing SAML attributes
+        :type attributes: Dict
+
+        :return: 2-tuple containing an optional name ID from the attribute list and
+            an attribute statement object
+        :rtype: Tuple[Optional[NameID, AttributeStatement]
+        """
+        name_id = None
         parsed_attributes = []
         attribute_names = {attribute.value: attribute for attribute in SAMLAttributes}
 
         for name, attribute_values in attributes.iteritems():
             if name in attribute_names:
                 name = attribute_names[name].name
 
-            parsed_attribute_values = self._parse_attribute_values(attribute_values)
+            name_id, parsed_attribute_values = self._parse_attribute_values(attribute_values)
```

Comment:
This will behave unpredictably if there are multiple items in `attributes`. `name_id` might even end up `None` because an attribute with a name ID shows up in `iteritems` before an attribute that lacks one.

Overall your strategy has been to just find _some_ name ID _somewhere, so I think it would be enough to assign to `name_id` only if the value coming out of `_parse_attribute_values` is None.

Reply:
I updated the code accordingly and added an additional tests covering the case with multiple SAML attributes

Reply:
I'm OK with this but the name is still unpredictable because the attributes ultimately come out of `attributes.iteritems()`. In Python 3 that order is predictable, in that it won't change over multiple runs, and I don't imagine this scenario happens a lot. My only real concern is that this will cause the unit tests to fail randomly.

Reply:
I can't imagine a case where it could fail. The code now overwrites NameID only in the case when it's not empty so it doesn't matter in what order we iterate attributes, i.e. it doesn't matter whether we get an attribute with NameID first or without